### PR TITLE
add section on mocks and recordings for Go

### PR DESCRIPTION
**Pull request by jhendrixMSFT**

### CODE REVIEW

File: docs/golang/implementation.md [hunk 1]
```diff
@@ -193,7 +193,55 @@ Let's take two examples:
 
 One of the key things we want to support is to allow consumers of the library to easily write repeatable unit-tests for their applications without activating a service. This allows them to reliably and quickly test their code without worrying about the vagaries of the underlying service implementation (including, for example, network conditions or service outages). Mocking is also helpful to simulate failures, edge cases, and hard to reproduce situations (for example: does code work on February 29th).
 
-{% include requirement/MUST id="golang-support-mocking" %} support mocking of network operations.
+{% include requirement/MUST id="golang-mock-interface-package" %} generate a sub-package containing interface definitions for all client operations.  The package name will be the same as the parent plus the `api` suffix.
```
Comment (JeffreyRichter) [inline docs/golang/implementation.md]: Does a competitor use the 'api' suffix?
Maybe we should use 'mock' suffix to indicate that this is for mocking. I think it's important to indicate this because we will not maintain backward compatibility if used. That is, we could add a new method to a FooClient (which is NOT breaking) but adding the method to the interface IS breaking. I wouldn't want to increment the module's major version if we add a new method to FooClient.

File: docs/golang/implementation.md [hunk 2]
```diff
@@ -193,7 +193,55 @@ Let's take two examples:
...
+
+{% include requirement/MUST id="golang-mock-interface-types" %} generate one interface type per client type that contains all of the client type's exported methods.  The interface type name will be the same as the client type name plus the `API` suffix.
```
Comment (JeffreyRichter) [inline docs/golang/implementation.md]: 'mock' suffix?

File: docs/golang/implementation.md [hunk 3]
```diff
@@ -193,7 +193,55 @@ Let's take two examples:
...
+
+{% include requirement/MUST id="golang-mock-lro-pages" %} generate interface types for LRO and pageable response types that contain all of the methods for their respective types.  The interface type name will be the same as the LRO/pageable response type name plus the `API` suffix.
```
Comment (JeffreyRichter) [inline docs/golang/implementation.md]: Remember that doing this requires that all these types (Client, LRO, pageable, etc.) do not have any fields; only methods. We'll need to add this to other parts of the guidelines.

File: docs/golang/implementation.md [hunk 4]
```diff
@@ -193,7 +193,55 @@ Let's take two examples:
...
+)
+
+// WidgetClientAPI contains the set of methods on the factory.WidgetClient type.
+type WidgetClientAPI interface {
+	CreateWidget(ctx context.Context, options factory.WidgetOptions) (*CreateWidgetOperation, error)
+	GetWidget(ctx context.Context, name string) (*factory.Widget, error)
+	ListWidgets(ctx context.Context, options factory.ListWidgetsOptions) (*ListWidgetsIterator, error)
+}
+
+var _ WidgetClientAPI = (*factory.WidgetClient)(nil)
```
Comment (JeffreyRichter) [inline docs/golang/implementation.md]: What's a bummer is that this compiles OK even if he interface is missing a method that exists on the WidgetClient type. I wish we had a way to look for this and prevent compilation.

File: docs/golang/implementation.md (hunk 1 above)
Comment (jhendrixMSFT) [reply docs/golang/implementation.md]: AWS and GCP use the suffix `iface`.  Would could change to match them, I stuck with `api` as that's what we do in track 1 so it didn't seem worth it to change.
Regarding breaking changes, it depends on how the interface is consumed.  If you embed it anonymously then no breaking change is introduced (this is what the whole `embedToIncludeNewMethods()` thing is in GCP's interface; it forces you do embed the interface anonymously).

File: docs/golang/implementation.md (hunk 4 above)
Comment (jhendrixMSFT) [reply docs/golang/implementation.md]: If we really care about this scenario I think we could have an `init()` function in this package that uses reflection to verify this.

File: docs/golang/implementation.md (hunk 1 above)
Comment (JeffreyRichter) [reply docs/golang/implementation.md]: I think this is still breaking at runtime due to the panic. 
Technically, if a method accepts an interface parameter and the interface changes to include a new method, then this is a breaking change. And, even if you use the embedToIncludeNewMethod() thing, then if the method calls the method, a runtime panic occurs. 
IMPORTANT: the guidelines need to say that an method that requires a Client, LRO, or Poller type as a parameter, need to require an interface with the actually-called methods instead so that a mocked type can be passed in.

File: docs/golang/implementation.md (hunk 2 above)
Comment (JeffreyRichter) [reply docs/golang/implementation.md]: I'm confused. Above you say that AWS/GCP use iface suffix but then here you say that AWS uses API (not iface) and GCP uses no suffix at all. Maybe you can include a link to the AWS/GCP examples?

File: docs/golang/implementation.md (hunk 4 above)
Comment (JeffreyRichter) [reply docs/golang/implementation.md]: Yes, good idea. We can ignore this for now and when we have the first actual problem of this type, we'll add the init()/reflection function.

File: docs/golang/implementation.md (hunk 2 above)
Comment (jhendrixMSFT) [reply docs/golang/implementation.md]: `iface` on the package name.  `API` on the interface type name in the case of AWS.

https://github.com/aws/aws-sdk-go/blob/master/service/amplify/amplifyiface/interface.go#L9
https://github.com/googleapis/google-cloud-go-testing/blob/master/bigquery/bqiface/interfaces.go#L15

Comment (jhendrixMSFT) [reply docs/golang/implementation.md]: I would be for changing the package name to use the `iface` suffix so we have consistency.

File: docs/golang/implementation.md (hunk 1 above)
Comment (jhendrixMSFT) [reply docs/golang/implementation.md]: It would only panic if you called it though; presumably this is a good thing, i.e. if you're calling it in your test framework then you likely want to mock it.
I will update the guidelines with the interface info described.

File: docs/golang/implementation.md (hunk 3 above)
Comment (jhendrixMSFT) [reply docs/golang/implementation.md]: Fixed

File: docs/golang/implementation.md (hunk 1 above)
Comment (jhendrixMSFT) [reply docs/golang/implementation.md]: Added section about accepting interfaces instead of concrete types.

File: docs/golang/implementation.md (hunk 2 above)
Comment (JeffreyRichter) [reply docs/golang/implementation.md]: OK

Comment (jhendrixMSFT) [reply docs/golang/implementation.md]: So change the package suffix to `iface` then?  How about the 'API` suffix on the type names?  I'm thinking we can drop that as the type names are already distinct due to being in different packages.

Comment (JeffreyRichter) [reply docs/golang/implementation.md]: Sure, drop the API suffix; add iface to the package name

Comment (jhendrixMSFT) [reply docs/golang/implementation.md]: Fixed

File: docs/golang/implementation.md (hunk 1 above)
Comment (jhendrixMSFT) [reply docs/golang/implementation.md]: Fixed
